### ISSUE TITLE
Use multiple threads to process inbound records

### ISSUE THREAD

**Issue by sophokles73** [eclipse/californium#216]
This is particularly problematic because Scandium (currently) uses a single thread only for processing all incoming messages.

This is a follow-up issue of #213

**Comment by sbernard31** [comment]
I know that the [commit](166cd6973133fd803d5e66157f12de270729bd76) referenced here, is not a PR and is probably not intended to be reviewed. But I see severals modifications on the same commit :
1) Usage of a new `RecordParsingException` instead of `HandshakeException`
2) Add a new utility class to create handshaker `Handshakers` and make hanshakers constructor not public
3) Some cleaning (remove comments , replacing \n by System.lineseperator, remove commented code ...)
4) Replace `GeneralSecurityException` by `RecordDecryptionException`
5) change log level ?
6) process handshake message asynchronously.

All of thoses changes, it's hard to review.

About 6), If I well understand, you authorize to handle several handshake messages for the same handshake at the same time and I'm afraid of the thread-safeness of this.

About #213, making incoming processing messages multi-threaded is probably the more complex problem I listed, but it is maybe not essential (just an optimization). I suppose we can mainly resolve this issue accessing PSKStore asynchronously. I can try to propose a PR, but I would like to integrate a recent version of Californium/Scandium in Leshan first.

**Comment by sophokles73** [comment]
As you recognized correctly, that branch is not meant as a PR. I was trying to better understand the challenges and what it would take to actually make some progress wrt the current problem but as well with the problem of only using a single thread for *all* messages (regardless of whether they are application or handshake).

> About 6), If I well understand, you authorize to handle several handshake messages for the same handshake at the same time and I'm afraid of the thread-safeness of this.

No, not at all. The intention is to make sure that at all times only a single (the next due one) handshake message is processed *per handshaker*. However, we can process messages for  different handshakes in parallel ...

>  I can try to propose a PR, but I would like to integrate a recent version of Californium/Scandium in Leshan first.

Be my guest to propose a PR for fixing PSKStore access in the meantime :-) BTW we seem to have a lot of problems with creating a *recent version* of Californium currently ...

**Comment by sbernard31** [comment]
> only a single (the next due one) handshake message is processed per handshaker.

Because of synchronized keyword on `doProcessMessage` ? I'm not sure it is enough, because this seems not to be the case for the rest of the code in the task (I think in particular about  `InboundMessageBuffer`)

**Comment by boaks** [comment]
I guess the idea of serialize the messages per handshaker is implemented in 

> public void run() {
> 	try {
> 		doProcessMessage(msg);
> 		onMessageProcessed(msg);
> 	} catch (RecordProcessingException e) {
> 		recordLayer.onRecordProcessingError(e);
> 	}
> }
> 

"process a message and afterwards schedule the processing of the next message of this handshake".
The question would be, if there is a situation, where `processMessage(Record record)` schedules the processing of a message in parallel. And that's not too easy to understand. The `nextReceiveSeq` may protect from more then one message being processed, but as the `inboundMessageBuffer` it currently seems to lack thread safety.

**Comment by sophokles73** [comment]
Guys,I appreciate your enthusiasm :-) but I would rather have you not review this branch yet. In fact, I have made lots of changes locally already which are not reflected in the branch ...
So far, this is only for me to explore possibilities and making some experiments.

**Comment by sophokles73** [comment]
Okay, I have created a PR addressing the issue. I have tried to keep the number of changes very low so that it is easier to review which I hereby encourage you guys to do :-)
